### PR TITLE
Preserves FIRRTL names

### PR DESCRIPTION
**Pull request by darthscsi**
This does a simple wire preservation.  This means:
1) nodes and wires become wires if they have names
2) names that don't start with '_' are preserved
3) names don't block constant propgation, but do block deleting wires.

### CODE REVIEW

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 1]
```diff
@@ -1612,6 +1607,7 @@ struct FIRStmtParser : public FIRParser {
   ParseResult parseWire();
   ParseResult parseRegister(unsigned regIndent);
 
+
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:1610]: Nit: stray newline.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 2]
```diff
@@ -3174,19 +3130,8 @@ ParseResult FIRStmtParser::parseNode() {
...
-  // If the node is named, then add a debug tap.
-  //
-  // TODO: Change this once the FIRRTL spec supports "named" vs. "unnamed"
-  // nodes.
-  if (isNamed)
-    builder.create<NodeOp>(
-        initializer.getType(), result, id, getConstants().emptyArrayAttr,
-        StringAttr::get(annotations.getContext(), modNameSpace.newName(id)));
+  //sym = getSymbolIfMissing(sym, id);
+  auto result = builder.create<NodeOp>(initializer.getType(), initializer, id,                                       annotations, sym);
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:3134]: Weird indentation (likely tripping `clang-format`).

File: test/Conversion/FIRRTLToHW/lower-to-hw.mlir [hunk 3]
```diff
@@ -205,9 +205,11 @@ firrtl.circuit "Simple"   attributes {annotations = [{class =
     // CHECK-NEXT: = comb.extract [[MOD2]] from 0 : (i8) -> i3
     %21 = firrtl.rem %3, %in3 : (!firrtl.sint<3>, !firrtl.sint<8>) -> !firrtl.sint<3>
 
-    // Nodes with names but no attribute are just dropped.
+    // Nodes with names become wires.
+    // CHECK-NEXT: %n1 = sv.wire
+    // CHECK-NEXT: sv.assign %n1, %in2
     %n1 = firrtl.node %in2  {name = "n1"} : !firrtl.uint<2>
-
+
```
Comment (seldridge) [inline test/Conversion/FIRRTLToHW/lower-to-hw.mlir:212]: Nit: extra whitespace characters.

**Review by seldridge — APPROVED** [review]
Let's give this a shot and see how it goes.

The codes has some whitespace issues which are likely running afoul of `clang-format`.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 4]
```diff
@@ -3174,19 +3130,8 @@ ParseResult FIRStmtParser::parseNode() {
...
-                                       annotations, sym);
-  // If the node is named, then add a debug tap.
-  //
-  // TODO: Change this once the FIRRTL spec supports "named" vs. "unnamed"
-  // nodes.
-  if (isNamed)
-    builder.create<NodeOp>(
-        initializer.getType(), result, id, getConstants().emptyArrayAttr,
-        StringAttr::get(annotations.getContext(), modNameSpace.newName(id)));
+  //sym = getSymbolIfMissing(sym, id);
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:3133]: Nit: probably can drop this commented-out line, same for below instance of same.

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp [hunk 5]
```diff
@@ -1470,12 +1471,20 @@ static LogicalResult canonicalizeSingleSetConnect(StrictConnectOp op,
...
 
-  // Remove the connect.
-  rewriter.eraseOp(op);
+    // Remove the connect
+    rewriter.eraseOp(op);
+  } else {
+    // bypass the decl, but keep it around
+    rewriter.startRootUpdate(connectedDecl);
+    connectedDecl->replaceAllUsesWith(ArrayRef{replacement});
+    op->setOperand(0, connectedDecl->getResult(0));
```
Comment (mikeurbach) [inline lib/Dialect/FIRRTL/FIRRTLFolds.cpp:1485]: The mutation on `connectedDecl` is being registered with the rewriter here, but this also mutates `op`. Does this need something like `rewriter.updateRootInPlace(op, [](){ op->setOperand(...) }` to be squeeky clean?

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp [hunk 6]
```diff
@@ -2304,7 +2304,11 @@ LogicalResult FIRRTLLowering::visitDecl(WireOp op) {
     symName = builder.getStringAttr(Twine("__") + moduleName + Twine("__") +
                                     name.getValue());
   }
-
+  if (!symName && name && !name.getValue().empty()) {
+    auto moduleName = cast<hw::HWModuleOp>(op->getParentOp()).getName();
+    symName = builder.getStringAttr(Twine("__") + moduleName + Twine("__") +
+                                    name.getValue());
```
[the remark this answers is not in the document]
Comment (mikeurbach) [reply lib/Conversion/FIRRTLToHW/LowerToHW.cpp:2310]: Might also be nice to pull out a helper method.

File: include/circt/Dialect/FIRRTL/FIRRTLOps.h [hunk 7]
```diff
@@ -30,6 +30,13 @@
 namespace circt {
 namespace firrtl {
 
+// is the name useless?
+bool isUselessName(circt::StringRef name);
+
+// works for regs, nodes, and wires
```
Comment (mikeurbach) [inline include/circt/Dialect/FIRRTL/FIRRTLOps.h:36]: style nit: could capitalize and punctuate.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp (hunk 4 above)
Comment (mikeurbach) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:3133]: Commented out uses and the definition of getSymbolIfMissing could be removed.

**Review by mikeurbach — APPROVED** [review]
Agree, let's go forward with this. Some minor nitpicks but looks good, and I've been trying it out to good effect.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 8]
```diff
@@ -1625,6 +1621,17 @@ struct FIRStmtParser : public FIRParser {
     return {};
   }
 
+  /// Remove the DontTouch annotation and return a valid symbol name if the
+  /// annotation exists. Ignore DontTouch that will apply only to a subfield.
+  StringAttr getSymbolIfMissing(StringAttr oldSym, StringRef id) {
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:1626]: Does this do what the comment says re:DontTouch annotations?

**Review by dtzSiFive — COMMENTED** [review]
Minor things noticed while looking it over :).  Thanks for tackling this.

Is there a place to document/announce/describe that our semantics are now that leading underscore in name means they can be deleted, otherwise they're preserved?

**Pull request merged**